x/accesscontrol/keeper/keeper.go

```diff
@@ -125,48 +126,60 @@ func (k Keeper) SetDependencyMappingDynamicFlag(ctx sdk.Context, messageKey type
+		return nil, err
+	}
+	return &dependencyMapping, nil
+}
+
+func (k Keeper) GetWasmDependencyMapping(ctx sdk.Context, contractAddress sdk.AccAddress, senderBech string, msgBody []byte, circularDepLookup ContractReferenceLookupMap) ([]*acltypes.WasmAccessOperation, error) {
 	// TODO: check the circularDepLookup
 	uniqueIdentifier := contractAddress.String()
 
 	if _, ok := circularDepLookup[uniqueIdentifier]; ok {
 		// we've already seen this contract, we should simply return synchronous access Ops
-		return types.SynchronousWasmDependencyMapping(contractAddress.String()), nil
+		return types.SynchronousWasmAccessOps(), nil
 	}
 	// add to our lookup so we know we've seen this identifier
 	circularDepLookup[uniqueIdentifier] = struct{}{}
 
-	store := ctx.KVStore(k.storeKey)
-	b := store.Get(types.GetWasmContractAddressKey(contractAddress))
-	if b == nil {
-		// return default (synchronous) dependency mapping so that wasm resource type as a whole
-		// won't have dynamic mapping disabled if already enabled
-		return acltypes.WasmDependencyMapping{
-			Enabled:   true, // if wasm resource type as a whole is disabled, this will be ignored anyway
-			AccessOps: types.SynchronousAccessOpsWithSelector(),
-		}, nil
+	dependencyMapping, err := k.GetRawWasmDependencyMapping(ctx, contractAddress)
+	if err != nil {
+		if err == sdkerrors.ErrKeyNotFound {
+			return types.SynchronousWasmAccessOps(), nil
+		}
+		return nil, err
 	}
-	dependencyMapping := acltypes.WasmDependencyMapping{}
-	k.cdc.MustUnmarshal(b, &dependencyMapping)
-	if dependencyMapping.Enabled && applySelector {
-		selectedAccessOps, err := k.BuildSelectorOps(ctx, contractAddress, dependencyMapping.AccessOps, senderBech, msgBody, circularDepLookup)
+
+	if dependencyMapping.Enabled {
+		// TODO: extend base access op with message-specific ops once message type identifier is passed in
+		selectedAccessOps, err := k.BuildSelectorOps(ctx, contractAddress, dependencyMapping.BaseAccessOps, senderBech, msgBody, circularDepLookup)
 		if err != nil {
-			return acltypes.WasmDependencyMapping{}, err
+			return nil, err
 		}
-		dependencyMapping.AccessOps = selectedAccessOps
+		return selectedAccessOps, nil
 	}
-	return dependencyMapping, nil
+
+	// use synchronous if dependency mapping is disabled
+	return types.SynchronousWasmAccessOps(), nil
 }
 
-func (k Keeper) BuildSelectorOps(ctx sdk.Context, contractAddr sdk.AccAddress, accessOps []acltypes.AccessOperationWithSelector, senderBech string, msgBody []byte, circularDepLookup ContractReferenceLookupMap) ([]acltypes.AccessOperationWithSelector, error) {
-	selectedAccessOps := []acltypes.AccessOperationWithSelector{}
+func (k Keeper) BuildSelectorOps(ctx sdk.Context, contractAddr sdk.AccAddress, accessOps []*acltypes.WasmAccessOperation, senderBech string, msgBody []byte, circularDepLookup ContractReferenceLookupMap) ([]*acltypes.WasmAccessOperation, error) {
+	selectedAccessOps := []*acltypes.WasmAccessOperation{}
```

Comment:
if we're restructuring this data type, should we make it a map[accessOperation]struct{} instead to basically be a set? we previously had a TODO to deduplicate access operations, and this might be a good way to go about it? I'm ok leaving this as is and addressing it later instead for simplicity

Reply:
yeah makes sense. Updated